packages/yew-router/src/history.rs

```diff
@@ -0,0 +1,237 @@
+                    for subscriber in subscribers {
+                        subscriber.emit(route.clone());
+                    }
+                })
+            } else {
+                None
+            }
+        }
+    }
+
+    fn update() {
+        if let Some(f) = Self::with(Self::update_inner) {
+            f();
+        }
+    }
+
+    fn current_route(maybe_state: &mut Option<Self>) -> Rc<Route> {
+        maybe_state.get_or_insert_with(Self::new).last_route.clone()
+    }
+
+    fn register(maybe_state: &mut Option<Self>, callback: Callback<Rc<Route>>) -> HistoryListener {
+        maybe_state
+            .get_or_insert_with(Self::new)
+            .subscribers
+            .push(callback.clone());
+        HistoryListener(callback)
+    }
+
+    fn unregister(maybe_state: &mut Option<Self>, callback: &Callback<Rc<Route>>) {
+        if let Some(state) = maybe_state {
+            if let Some(index) = state
+                .subscribers
+                .iter()
+                .position(|subscriber| subscriber == callback)
+            {
+                state.subscribers.remove(index);
+            }
+        }
+    }
+}
+
+pub struct HistoryListener(Callback<Rc<Route>>);
+
+impl Drop for HistoryListener {
+    fn drop(&mut self) {
+        HistoryState::with(|state| HistoryState::unregister(state, &self.0));
+    }
+}
+
+pub fn dispatch(action: HistoryAction) {
+    let history = window().history().expect("no history");
+    match action {
+        HistoryAction::Push(mut route) => {
+            route.apply_base();
+            history
+                .push_state_with_url(&JsValue::NULL, "", Some(&route.url()))
+                .expect("push history");
+
+            // Not triggered automatically by `pushState`.
+            HistoryState::update();
+        }
+        HistoryAction::Replace(mut route) => {
+            route.apply_base();
+            history
+                .replace_state_with_url(&JsValue::NULL, "", Some(&route.url()))
+                .expect("replace history");
+
+            // Not triggered automatically by `replaceState`.
+            HistoryState::update();
+        }
+        HistoryAction::Back => history.back().expect("back history"),
+        HistoryAction::Forward => history.forward().expect("forward history"),
+        HistoryAction::Go(index) => history.go_with_delta(index).expect("go history"),
+    }
+}
+pub fn push(route: Route) {
+    dispatch(HistoryAction::Push(route));
+}
+pub fn replace(route: Route) {
+    dispatch(HistoryAction::Replace(route));
+}
+pub fn forward() {
+    dispatch(HistoryAction::Forward);
+}
+pub fn back() {
+    dispatch(HistoryAction::Back);
+}
+pub fn go(index: i32) {
+    dispatch(HistoryAction::Go(index));
+}
+pub fn current() -> Rc<Route> {
+    HistoryState::with(HistoryState::current_route)
+}
+pub fn register(callback: Callback<Rc<Route>>) -> HistoryListener {
+    HistoryState::with(|state| HistoryState::register(state, callback))
+}
```

Comment:
Why are these functions (and `dispatch`) `pub`? The Router API should be used for these operations by the user. Let's make these functions `pub(crate)`.

Reply:
I can make the whole `history` API private if you want to be conservative, but it does seem useful to be able to say go "forward", "back", etc, don't you think?

Reply:
How about we move it to [`gloo`](https://github.com/rustwasm/gloo/) so not only this crate but others can also benefit from this abstraction? I've done the same for `yew-services` as part of my refactor PR.

Reply:
It would be a bit awkward since this API uses the `Route` type, so `Route` would have to be moved into `gloo`, and then that means it will need a public API. I'll make it private for now and then we can decide what to do later.

Reply:
I'd still move everything except the `Route` type to gloo. That is, all the wrappers for the history API.

Reply:
Let's leave those as `pub(crate)` for now. I'll see if it's something that can be make agnostic of this crate (hopefully) as part of refactor PRs.